File: 2020/2020W33/preppindataw33.R
library(dplyr)
library(readxl)
library(lubridate)
library(tidyr)

xlsx <- "F:/Data/2020W33.xlsx"

final <- read_excel(xlsx, sheet = 'Daily Sales', col_types = c("date", 
                                                               "text", "numeric")) %>%
  merge(read_excel(xlsx, sheet = 'Scent'), by='Scent Code') %>%
  mutate('Week Start' = floor_date(`Date` %m+% days(-3), unit='week') %m+% days(3),
         'Units Sold' = `Daily Sales`/`Price`) %>%
  group_by(`Week Start`, `Scent`) %>%
  summarise('Cost' = max(`Cost`),
            'Weekly Units Sold' = sum(`Units Sold`),
            'Weekly Sales' = sum(`Daily Sales`)) %>%
  merge(read_excel(xlsx, sheet = 'Orders'), by.x='Week Start', by.y='Date', all.x=TRUE) %>%
  mutate_at(vars('Units Ordered'), ~replace_na(., 0)) %>%
  mutate('Waste' = `Units Ordered` - `Weekly Units Sold`,
         'Waste Cost' = `Waste`*`Cost`,
         'Profit' = `Weekly Sales`-`Waste Cost`) %>%
  group_by(`Scent`) %>%
  summarise('Total Profit' = sum(`Profit`)) %>%
  mutate('Profitability Rank' = rank(desc(`Total Profit`)))
  
View(final)

File: 2020/2020W33/PreppindataW33.py
import pandas as pd
from datetime import timedelta

MON, TUE, WED, THU, FRI, SAT, SUN = range(7)

def get_weekstart(xdate, startwd):
    offset = (xdate.weekday()-startwd)%7
    return xdate - timedelta(days=offset)

xlsx = pd.ExcelFile("F:/Data/2020W33.xlsx")
order = pd.read_excel(xlsx,'Orders')   
prod= pd.read_excel(xlsx,'Scent')   
sales = pd.read_excel(xlsx,'Daily Sales')  

final = pd.merge(sales,prod, on='Scent Code')
final['Units Sold'] = final['Daily Sales']/final['Price'] 
final['Week Start'] = final['Date'].apply(lambda x: get_weekstart(x,WED)) 
final = final.groupby(['Week Start','Scent'], as_index=False).agg({'Daily Sales':'sum', 'Units Sold':'sum', 'Cost':'max'})
final = final.merge(order, how='left', left_on='Week Start', right_on='Date').fillna(0)
final['Total Profit'] = final['Daily Sales'] - (final['Units Ordered'] - final['Units Sold'])*final['Cost']

final = final.groupby(['Scent'], as_index=False).agg({'Total Profit':'sum'})
final['Profitability Rank'] = final['Total Profit'].rank(ascending=False)
final = final[['Profitability Rank', 'Scent', 'Total Profit']]
